File: class10/java/PES1001-UserStoryParser.java
/*
 * PES1001 - Parser/Validador de User Stories
 * Aula 10: Levantamento e Elicitação de Requisitos
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Valida user stories no formato "Como [ator], quero [ação] para [benefício]"
 * e extrai os três componentes.
 *
 * Compilar/executar:
 *   javac PES1001-UserStoryParser.java && java UserStoryParser
 */

import java.util.*;
import java.util.regex.*;

class UserStoryParser {

    static final Pattern USER_STORY_PATTERN = Pattern.compile(
        "^Como\\s+(.+?),\\s*quero\\s+(.+?),\\s*para\\s+(.+?)$",
        Pattern.CASE_INSENSITIVE
    );

    static class Componentes {
        String ator, acao, beneficio;

        Componentes(String ator, String acao, String beneficio) {
            this.ator = ator;
            this.acao = acao;
            this.beneficio = beneficio;
        }
    }

    static class Analise {
        boolean valida;
        Componentes componentes;
        List<String> erros = new ArrayList<>();
        String erroGeral;

        static Analise invalida(String erro) {
            Analise a = new Analise();
            a.valida = false;
            a.erroGeral = erro;
            return a;
        }

        static Analise valida(Componentes c) {
            Analise a = new Analise();
            a.valida = true;
            a.componentes = c;
            return a;
        }
    }

    public static Componentes validarUserStory(String texto) {
        Matcher m = USER_STORY_PATTERN.matcher(texto.trim());
        if (!m.matches()) return null;
        String beneficio = m.group(3).trim();
        if (beneficio.endsWith(".")) beneficio = beneficio.substring(0, beneficio.length() - 1);
        return new Componentes(m.group(1).trim(), m.group(2).trim(), beneficio);
    }

    public static Analise analisarUserStory(String texto) {
        Componentes c = validarUserStory(texto);
        if (c == null) {
            return Analise.invalida("Formato inválido. Use: Como [ator], quero [ação] para [benefício]");
        }
        Analise a = Analise.valida(c);
        if (c.acao.split("\\s+").length < 2) {
            a.erros.add("Ação muito curta — seja mais específico(a)");
        }
        if (c.beneficio.split("\\s+").length < 2) {
            a.erros.add("Benefício muito vago — explique o valor entregue");
        }
        return a;
    }

    public static void main(String[] args) {
        String[] stories = {
            "Como cliente, quero buscar produtos por nome, para encontrar rapidamente o que desejo comprar",
            "Como admin, quero listar usuários, para gerenciar acessos",
            "Como usuário, quero login para segurança",
            "Cliente quer buscar produtos",
        };

        System.out.println("=".repeat(60));
        System.out.println("  VALIDADOR DE USER STORIES");
        System.out.println("=".repeat(60));

        for (int i = 0; i < stories.length; i++) {
            System.out.printf("%n--- Story %d ---%n", i + 1);
            System.out.printf("  Texto: \"%s\"%n", stories[i]);
            Analise analise = analisarUserStory(stories[i]);
            if (analise.valida) {
                Componentes c = analise.componentes;
                String qualidade = analise.erros.isEmpty() ? "Boa" : "Revisar";
                System.out.printf("  ✓ Válida (%s)%n", qualidade);
                System.out.printf("    Ator: %s%n", c.ator);
                System.out.printf("    Ação: %s%n", c.acao);
                System.out.printf("    Benefício: %s%n", c.beneficio);
                for (String e : analise.erros) {
                    System.out.printf("    ⚠️  %s%n", e);
                }
            } else {
                System.out.printf("  ✗ Inválida: %s%n", analise.erroGeral);
            }
        }
    }
}

File: class10/python/PES1001-UserStoryParser.py
"""
PES1001 - Parser/Validador de User Stories
Aula 10: Levantamento e Elicitação de Requisitos
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Valida user stories no formato "Como [ator], quero [ação] para [benefício]"
e extrai os três componentes.
"""

import re

USER_STORY_PATTERN = re.compile(
    r"^Como\s+(.+?),\s*quero\s+(.+?),\s*para\s+(.+?)$",
    re.IGNORECASE
)


def validar_user_story(texto):
    """Valida e extrai os componentes de uma user story."""
    texto = texto.strip()
    match = USER_STORY_PATTERN.match(texto)
    if not match:
        return None
    ator = match.group(1).strip()
    acao = match.group(2).strip()
    beneficio = match.group(3).strip().rstrip(".")
    return {"ator": ator, "acao": acao, "beneficio": beneficio}


def analisar_user_story(texto):
    """Analisa uma user story fornecendo feedback."""
    resultado = validar_user_story(texto)
    if resultado is None:
        return {"valida": False, "erro": "Formato inválido. Use: Como [ator], quero [ação] para [benefício]"}
    erros = []
    if len(resultado["acao"].split()) < 2:
        erros.append("Ação muito curta — seja mais específico(a)")
    if len(resultado["beneficio"].split()) < 2:
        erros.append("Benefício muito vago — explique o valor entregue")
    return {
        "valida": True,
        "componentes": resultado,
        "erros": erros,
        "qualidade": "Boa" if not erros else "Revisar"
    }


if __name__ == "__main__":
    stories = [
        "Como cliente, quero buscar produtos por nome, para encontrar rapidamente o que desejo comprar",
        "Como admin, quero listar usuários, para gerenciar acessos",
        "Como usuário, quero login para segurança",  # benefício curto (alerta)
        "Cliente quer buscar produtos",              # formato inválido
    ]

    print("=" * 60)
    print("  VALIDADOR DE USER STORIES")
    print("=" * 60)

    for i, story in enumerate(stories, 1):
        print(f"\n--- Story {i} ---")
        print(f"  Texto: \"{story}\"")
        analise = analisar_user_story(story)
        if analise["valida"]:
            c = analise["componentes"]
            print(f"  ✓ Válida ({analise['qualidade']})")
            print(f"    Ator: {c['ator']}")
            print(f"    Ação: {c['acao']}")
            print(f"    Benefício: {c['beneficio']}")
            for e in analise["erros"]:
                print(f"    ⚠️  {e}")
        else:
            print(f"  ✗ Inválida: {analise['erro']}")
